### PR TITLE
Break reference cycle in AccumulateGrad::variable

### LINKED ISSUE #3818: Memory leak when doing backward with grad as yourself
@zdevito  You said you wanted to make a python repro?

```
void backward(std::vector<Variable> outputs) {
   variable_list varlst;
   function_list funclst;
   for (auto v : outputs) {
     funclst.emplace_back(v.grad_fn(), v.output_nr());
     varlst.emplace_back(v);
   }
   detail::engine.execute(funclst, varlst, false);
}
```

This will leak a bunch of Tensors and in general be very sad for whatever Variables used.

### PR DESCRIPTION
**Pull request by skrah**
Fixes #3818.

### CODE REVIEW

**Comment by skrah** [comment]
For the description of the cycle, it is best to look at @colesbury's Python example in #3818 or at the test case in this PR, which is based on the example.

As @colesbury has mentioned, the cycle occurs after `AccumulateGrad::variable` has been changed from a weak reference to an actual `Variable` in 1290e586.

I've verified that the cycle does not occur in the previous revision 820143f4.

In theory there are several locations where the cycle could be broken, but I think `variable.grad` or `variable.grad_fn` are not feasible, because they're assignment targets of fresh shared pointers in various places.

So that leaves `AccumulateGrad::variable`. Since `Variable` is no longer a `std::shared_ptr`, the only way I could think of is to introduce a new `WeakVariable` that is a minimal container for `c10::weak_intrusive_ptr`.

Using the weak variable, this PR restores the tricky `AccumulateGrad::apply` behavior from 820143f4.  


The leaks are fixed and I can see no performance degradation in various test cases.


@ezyang @colesbury I think the PR is at a stage where you can decide if this approach has a chance (tests pass).

If so, I can audit the use of `weak::intrusive_ptr` a bit more, run Valgrind on the tests etc.

**Comment by ezyang** [comment]
cc @zheng-xq who is on the record for wanting to kill weak pointer support on `intrusive_ptr`.

File: torch/csrc/autograd/functions/accumulate_grad.h [hunk 1]
```diff
@@ -7,11 +7,12 @@
...
 
 struct TORCH_API AccumulateGrad : public Node {
-  explicit AccumulateGrad(Variable variable_);
+  explicit AccumulateGrad(const Variable& variable_);
 
   variable_list apply(variable_list&& grads) override;
 
-  Variable variable;
+  WeakVariable variable;
+  WeakVariable variable_grad;
```
Comment (ezyang) [inline torch/csrc/autograd/functions/accumulate_grad.h]: About here would be a good explanation about why it's OK for this to be weak. (Not too hard to understand: if the variable is dead, no need to accumulate into it.) But why also record `variable_grad`? That would be a good thing to explain in line here.

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 2]
```diff
@@ -28,17 +29,37 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
...
+
+  auto var = variable.lock();
+  // It's possible that the Variable went out of scope and was freed.
+  // We still need to handle the unlikely case of someone holding to its grad.
+  if (!var.defined()) {
+    auto var_grad = variable_grad.lock();
+    // Everything was freed. Nothing to do.
+    if (!var_grad.defined()) return variable_list();
+    // Now here's the hard part. If both the new_grad and var_grad require grad
+    // then we just acumulate the data in place (as we'd do if the Variable was
```
Comment (ezyang) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: nit: accumulate

**Comment by ezyang** [comment]
For reference, the old code is https://github.com/pytorch/pytorch/blob/820143f4af0128dad56d6a610354e8a79510d877/torch/csrc/autograd/functions/accumulate_grad.cpp#L41

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 3]
```diff
@@ -28,17 +29,37 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
...
+    // alive). Otherwise, we'd need to perform the out-of-place reduction, but
+    // since the user only holds a reference to .grad and there's no way to
+    // give him the new Value, we just assume that they know these attributes
+    // are changing when using higher order graphs.
+    if (GradMode::is_enabled() && var_grad.requires_grad() && grads[0].requires_grad()) {
```
Comment (ezyang) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: I don't think the translation of `volatile()` to `!requires_grad()` is valid here. We can see that there is trouble because if we read the comment, we see that the condition here doesn't actually match the logic that would have been executed if the variable was alive.

This code in-place updates:

```
import torch
W = torch.randn(10, requires_grad=True)
W.grad = torch.zeros(10)
old_grad = W.grad
x =torch.randn(10, requires_grad=True)
z = W * x
z.backward(torch.ones(z.size()))
print(old_grad)
```

But after your change it will not (because `W.grad` does not require grad)

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 4]
```diff
@@ -52,10 +73,16 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
       // If the function has post hooks (for example, a DDP allreduce hook),
       // call_function in Engine.cpp will temporarily bump the refcount by one, hence the
       // addition of !post_hooks().empty().
-      variable.grad() = new_grad.detach();
+      var.grad() = new_grad.detach();
     } else {
-      variable.grad() = new_grad.clone();
+      var.grad() = new_grad.clone();
     }
+    variable_grad = WeakVariable(var.grad()); // We need to update our reference
```
Comment (ezyang) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: Importantly, there is only ever one grad accumulator for a variable.

**Review by ezyang — CHANGES_REQUESTED** [review]
`requires_grad` handling is wrong, I think

File: torch/csrc/autograd/functions/accumulate_grad.cpp (hunk 4 above)
Comment (skrah) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: Agreed, actually I didn't quite understand how the rest of the comment applies to this line.  I now think the rest is for the next line, so I'll remove it.

File: torch/csrc/autograd/functions/accumulate_grad.cpp (hunk 3 above)
Comment (skrah) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: Yes, it does not look right.  Apparently the current code translates

```
 } else if (var->grad->is_volatile) {
    acc_inplace(var->grad, new_grad);
```

to

```
  } else if (!GradMode::is_enabled()) {
...
```

So the latest version of this PR just uses the same code as below that don't need `variable` to be defined.  I've also added a test for `sparse`.

**Comment by skrah** [comment]
@ezyang I think I'm a bit stuck on the `detached` test case, I wonder if the following is a bug or a feature:

```
import torch
import gc

W = torch.arange(10.0, requires_grad=True)
W.grad = torch.zeros(10)

# saved_grad = W.grad
detached_grad = W.grad.detach()

x = torch.arange(10.0, requires_grad=True)
z = W + x

del W
gc.collect()


z.backward(torch.ones(z.size()))

# print(saved_grad)
print(detached_grad)
```

`master` updates `detached_grad`, this PR does not (yet).  However, if you uncomment the `saved_grad` lines, this PR updates both `detached_grad` and `saved_grad`.

I'm not sure which is correct. If only detached references exist, should the detached reference be updated?

File: torch/csrc/autograd/functions/accumulate_grad.h (hunk 1 above)
Comment (skrah) [reply torch/csrc/autograd/functions/accumulate_grad.h]: Yes, definitely, I'll add more comments (also in other places) when the `detach()` question is settled.

**Comment by ezyang** [comment]
cc @yf225

**Comment by ezyang** [comment]
My gut feeling is that you're asking the wrong question. Detach or not, if I do an out-of-place "update" to a variable, any other references to that variable will not see the update; only an in-place update will be seen. What your example seems to be saying, is that you are switching between doing an in-place update and an out-of-place update depending on whether or not some variables are live. Are you sure this is what you want?

(Bug me again on Monday if you want me to try to actually figure out what should happen here. The above is just some half-baked thoughts.)

**Comment by skrah** [comment]
> What your example seems to be saying, is that you are switching between doing an in-place update and an out-of-place update depending on whether or not some variables are live. Are you sure this is what you want?

No, not really. :)  But  820143f (which I'm trying to emulate), apparently has the same behavior:

```python
from torch.autograd import Variable
from torch import arange, zeros, ones
import torch
import gc

W = Variable(arange(0.0, 2.0), requires_grad=True)
W.grad = Variable(zeros(2))

# saved_grad = W.grad
# saved_grad.volatile=True
detached_grad = W.grad.detach()

x = Variable(arange(0.0, 2.0), requires_grad=True)

z = W + x
del W
gc.collect()
z.backward(ones(z.size()))

# print(saved_grad)
print(detached_grad)
```

I'm not working on `pytorch` this week (and probably next), perhaps I'll have some inspiration in the meantime.
